base/server/healthcheck/pki/server/healthcheck/certs/expiration.py

```diff
@@ -0,0 +1,163 @@
+# Authors:
+#     Dinesh Prasanth M K <dmoluguw@redhat.com>
+#
+# Copyright Red Hat, Inc.
+#
+# SPDX-License-Identifier: GPL-2.0-or-later
+#
+import logging
+import time
+
+from datetime import datetime
+
+from pki.server.healthcheck.certs.plugin import CertsPlugin, registry
+from ipahealthcheck.core.plugin import Result, duration
+from ipahealthcheck.core import constants
+
+logger = logging.getLogger(__name__)
+
+
+@registry
+class CASystemCertExpiryCheck(CertsPlugin):
+    """
+    Check the expiry of CA's system certs
+    """
+
+    @duration
+    def check(self):
+
+        if not self.instance.exists():
+            logger.debug('Invalid instance: %s', self.instance.name)
+            yield Result(self, constants.CRITICAL,
+                         msg='Invalid PKI instance: %s' % self.instance.name)
+            return
+
+        self.instance.load()
+
+        ca = self.instance.get_subsystem('ca')
+
+        if not ca:
+            logger.info("No CA configured, skipping CA System Cert Expiry check")
+            return
+
+        certs = ca.find_system_certs()
+
+        for cert in certs:
+            # Get current server date in millis and convert it into timestamp
+            current_date = datetime.fromtimestamp(int(round(time.time())))
+
+            # Cert's expiry date is returned in millis. Convert to timestamp
+            expiry_date = datetime.fromtimestamp(cert.get('not_after') / 1000.0)
+
+            # Throw an error if expiry date cannot be retrieved
+            if expiry_date is None:
+                logger.error("Expiry date of %s is None", cert['nickname'])
+                yield Result(self, constants.ERROR,
+                             nickname=cert['nickname'],
+                             cert_id=cert['id'],
+                             msg='Unable to get cert\'s expiry date')
+
+                continue
+
+            expiry_date_human = expiry_date.strftime("%b %d %Y")
+
+            # Check if expired
+            delta = (expiry_date - current_date).days
+            if delta < 0:
```

Comment:
So if delta is 0, might we want to check if this one is expired as well? I guess a cert can be expired if the time has passed during a single day.

Reply:
OH, I guess this might depend upon how that days of the delta is evaluated. For instance if (expiry_date - current_date) = something small like -100 seconds..... What does the ".days" property give you? 0 or -1 ???

Reply:
Do you want me to change this to `<=` ? Note that we are converting millis to days ( as we do elsewhere). So, the "time" in the actual cert and the calculated time were little off (around 3 hours)

Reply:
>  What does the ".days" property give you? 0 or -1 ???

It should give 0... Note, I'm comparing only `days`. So, time doesn't matter.

Reply:
Maybe just do the compare of the raw timestamps before converting to days. Thus if the number is less than 0 we know for sure if it's expired and > than 0, still valid, if just barely.  In this scenario in the highly unlikely case that it is exactly zero, just treat it as expired. In my opinion. Then then the days can be used just to detect the impending expiration scenario..